### PR TITLE
[SPARK-47316][SQL] Fix TimestampNTZ in Postgres Array

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

For Postgres, TimestampNTZ works well for plain TimestampNTZ types but not for nested ones, typically for now: array.

This PR makes it work.

### Why are the changes needed?

feature and data consistency 

### Does this PR introduce _any_ user-facing change?

preferTimestampNTZ jdbc option now works for array([timestamp|time|...]) in postgres 
### How was this patch tested?

new tests
### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -87,17 +87,26 @@ abstract class JdbcDialect extends Serializable with Logging {
...
    * @return The actual DataType (subclasses of [[org.apache.spark.sql.types.DataType]])
    *         or null if the default type mapping should be used.
    */
   def getCatalystType(
-    sqlType: Int, typeName: String, size: Int, md: MetadataBuilder): Option[DataType] = None
+      sqlType: Int,
+      typeName: String,
+      size: Int,
+      md: MetadataBuilder,
+      isTimestampNTZ: Boolean): Option[DataType] = getCatalystType(sqlType, typeName, size, md)
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: It is not a breaking change. It has a default implementation which fallbacks to the original method.

Carrying in meta works but is not obvious for third-party developers.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: I feel this is not a good API design. Are we going to keep deprecating and adding new overloads with extra parameters every time we want to pass more information? AFAIK we are already using `MetadataBuilder` to carry decimal precision information.

[the remark this answers is not in the document]
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:109]: Indeed, the precision is in the method parameters, and the scale is in the meta:)

[the remark this answers is not in the document]
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala]: Comments addressed, please take a look again when you are available

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master.

Thank you very much @cloud-fan @dongjoon-hyun

**Pull request closed without merging**